src/components/Presenter.js

```diff
@@ -0,0 +1,358 @@
+    const next = this.loadNextItem(PresenterDirection.LEFT);
+    if (this.timerChange) {
+      clearTimeout(this.timerChange);
+      this.timerChange = null;
+    }
+
+    this.setState({
+      index: next.index,
+      direction: PresenterDirection.LEFT,
+      current: {
+        item: next.item,
+        status: ItemStatus.LOADED,
+      },
+    });
+  }
+
+  loadNextItem(dir = null) {
+    const { current, index, direction } = this.state;
+    const { items } = this.props;
+
+    var nidx = index;
+
+    const d = dir ? dir : direction;
+
+    if (d === PresenterDirection.RIGHT) {
+      nidx = index + 1 < items.length ? index + 1 : 0;
+    } else {
+      nidx = index - 1 >= 0 ? index - 1 : items.length - 1;
+    }
+
+    return {
+      index: nidx,
+      item: items[nidx],
+    };
+  }
+
+  handleEnd() {
+    this.setState({
+      current: {
+        ...this.state.current,
+        status: ItemStatus.LEAVING,
+      },
+    });
+  }
+
+  trickyStart() {
+    return setTimeout(() => {
+      this.setState({
+        current: {
+          ...this.state.current,
+          status: this.state.current.status + 1,
+        },
+      });
+    }, 20);
+  }
+
+  handlePausePlay() {
+    const { timeout } = this.props;
+    if (this.timerChange) {
+      clearTimeout(this.timerChange);
+      this.timerChange = null;
+    }
+
+    const { status } = this.state;
+
+    if (status === PresenterStatus.PLAYING) {
+      this.setState({
+        status: PresenterStatus.PAUSED,
+        current: {
+          ...this.state.current,
+          status: ItemStatus.ENTERED,
+        },
+      });
+    } else if (status === PresenterStatus.PAUSED) {
+      this.setState({
+        status: PresenterStatus.PLAYING,
+      });
+      if (this.timerChange) {
+        clearTimeout(this.timerChange);
+      }
+      this.timerChange = setTimeout(this.handleEnd, timeout);
+    }
+  }
+
+  trickyEnd() {
+    return setTimeout(() => {
+      const next = this.loadNextItem();
+      this.state.current.node.removeEventListener('transitionend', this.handleNodeAnimEnd);
+      this.setState({
+        index: next.index,
+        current: {
+          item: next.item,
+          status: ItemStatus.LOADED,
+        },
+      });
+    }, 20);
+  }
+
+  setupNode(node) {
+    const { current } = this.state;
+    if (node && !current.node && current.status === ItemStatus.LOADED) {
+      const DOMNode = ReactDOM.findDOMNode(node);
+      this.setState(
+        {
+          current: {
+            ...this.state.current,
+            node: DOMNode,
+          },
+        },
+        () => {
+          this.trickyStart();
+        }
+      );
+
+      DOMNode.addEventListener('transitionend', this.handleNodeAnimEnd, false);
+    }
+  }
+
+  handleNodeAnimEnd(e) {
+    const { current: { node, status } } = this.state;
+    if (status === ItemStatus.ENTERING || status === ItemStatus.LEAVING) {
+      const { timeout } = this.props;
+      const ns = status + 1;
```

Comment:
Please use meaningful/readable names.
Instead of `ns` you can use `newStatus` if it stands for...